### PR TITLE
Add noise map generation utilities

### PR DESCRIPTION
**Pull request by tomrijnbeek**
## ✨ What's this?
This PR introduces a new namespace that can be used to automatically generate noise. Specifically, we expose an interface that can return a value for a rectangular area based on an underlying noise function. This PR includes basic uniform noise and Perlin noise (the latter implementation pretty much straight up copied from TD).

## 🔍 Why do we want this?
Any algorithm generating 2D randomness (whether graphically or mechanically) will need functionality similar to the functionality in this PR. While uniform randomness is fairly simple to generate, Perlin randomness is a bit fiddly. Having reusable methods to "just" give you a Perlin noise map sounds hugely beneficial to anybody bootstrapping some of this generation in their game.

## 🏗 How is it done?
This PR describes an extendable set of interfaces for generating noise. The result of generating noise is an interface that allows the caller to look up a noise value for any coordinate within a rectangular area (upper bounds exclusive). This has already proven beneficial since Perlin noise and uniform noise have different methods of getting to a value based on their underlying noise array. It also allows for extendibility such as for example transformation methods, combining multiple noise maps in interesting ways, etc. without having to have to deal with a different interface. Of course performance concerns creep in at that point, but the current interface doesn't preclude you to "pre-calculate" a noise texture by sampling the underlying noise functions, though it is unlikely to give a huge performance boost overall unless you sample the same points a lot.

As explained above, the current PR introduces two noise types:
* Uniform noise, which creates a "noise texture" with pseudo-randomly generated values for each "pixel". It also allows you to specify an interpolation method that is used to handle values between pixels. Nearest and bilinear are the most logical here (bilinear wraps around).
* Perlin noise. This implementation is copied from TD, and is generated by using a lattice of vectors. The noise value at a given point is generated from some smart interpolation of dot products to create an overall smoother noise.

In addition, since the noise functionality had a need to pass in an interpolation method, I created a structure where interpolation methods have a very enum-like behaviour, but are in fact static objects under the hood. Introducing further interpolation methods should not be too hard, though further design will have to be done if we want to do something like (bi)cubic or Bezier interpolation, since they need additional values besides "from" and "to".

### 🔬 Why not another way?
There are a few decisions I made along the way. These are the options I remember rejecting:

* Make noise generation generic. Wouldn't it be great if you could generate a noise of floats, doubles, even vectors? While yes, it would be great, I don't think one noise generation algorithm maps to another domain trivially, so sticking with a simple value seems like it made the most sense. You can also transform the double into something else if really needed. All in all, I believe the simplicity in this interface outweighs hypothetical use cases.
* Make interpolation generic. Same as above: applying interpolation to colours and vectors can be useful, though it isn't immediately clear you would use exactly the same techniques. On top of that, it just sounds like a pain if you have to apply for each interpolation type how to add, multiply, etc. given that there just isn't a way in C# to add type restrictions based on the present operators. All in all, I followed the same argument as above: let's stick with something simple that works, before trying to come up with hypothetical use cases. If we need a generic interpolation framework, we can revisit when we have a solid use case.
* More tests. One would be right to argue that the current PR has very few tests. I could definitely extensively test both the interpolation and noise generation, but instead I opted to only test some basic properties of the noise generation. The reason is only partly laziness: I think property tests are super helpful for this domain, but the documentation of FsCheck is incredibly hard to delve through given that the majority of examples is in F# and doesn't seem to easily map to C# + XUnit. I'd much rather find a property testing framework that is not F# based, or otherwise find a better method of setting up these tests.

### 🦋 Side effects
We may want to consider moving the existing `Interpolate` methods over, but since they are floats, and the new interpolation methods work with doubles, this would potentially be a bigger change.

I can think of many additional improvements. I have created issues for all of them (I might remember more later): #222, #223, #224

## 💡 Review hints
I am aware this PR probably will not be without controversial choices, even if the actual added functionality is not ground-breaking. I considered splitting it already, but given GitHub's awkward support on stacking PRs and the fact that all of this goes together somewhat, I decided to keep it.

Feel free to do an initial pass over the proposed interfaces and APIs first before delving in implementations.

I would ask to focus review comments on the overall structure and decisions made here, rather than on the choice of what to include or not. As seen above, it was easy to come up with many future improvements to this framework, including potential performance optimisations. My hope with this PR is to have an initial foundation submitted.

### CODE REVIEW

File: Bearded.Utilities/Noise/NoiseMap.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+using System;
+
+namespace Bearded.Utilities.Noise
+{
+    public static class NoiseMap
+    {
+        public static INoiseMap FromArray(double[,] array, IInterpolationMethod2 interpolation) =>
```
Comment (amulware) [inline Bearded.Utilities/Noise/NoiseMap.cs:7]: This really makes me question the name even more, sounds more like this is a ContinuousMap (or maybe not even a Map?), not a NoiseMap.

File: Bearded.Utilities/Noise/NoiseMap.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                x *= width;
+                y *= height;
+
+                // The noise array keeps track of the values at the center of each "pixel", so for the interpolation, we
+                // need to move the array coordinates by (0.5, 0.5), which is the same as moving the x, y by
+                // (-0.5, -0.5).
+                x -= 0.5;
+                y -= 0.5;
+
+                var xBelow = MoreMath.FloorToInt(x);
```
Comment (amulware) [inline Bearded.Utilities/Noise/NoiseMap.cs:46]: This can easily end up as -1 if x is passed in as 0, right? that doesn't seem right?

File: Bearded.Utilities/Noise/NoiseMap.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                var xAbove = MoreMath.CeilToInt(x);
+                var yBelow = MoreMath.FloorToInt(y);
+                var yAbove = MoreMath.CeilToInt(y);
+                var u = x - xBelow;
+                var v = y - yBelow;
+
+                xBelow = (xBelow + width) % width;
```
Comment (amulware) [inline Bearded.Utilities/Noise/NoiseMap.cs:53]: So this actually wraps around? Is that really intended? I definitely didn't expect it. I would have thought it would map 0 to the first pixel and 1 to the last, instead both map to the average of first and last?

I'm not sure if wrapping is good default behaviour. If anything maybe we have to inject it, if we want to support it?

File: Bearded.Utilities/Noise/NoiseMap.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        private sealed class ArrayNoiseMap : INoiseMap
+        {
+            private readonly double[,] array;
+            private readonly IInterpolationMethod2 interpolation;
+            private readonly int width;
+            private readonly int height;
+
+            public ArrayNoiseMap(double[,] array, IInterpolationMethod2 interpolation)
+            {
+                this.array = array;
```
Comment (amulware) [inline Bearded.Utilities/Noise/NoiseMap.cs:19]: Is it intended that we don't make a defensive copy? I.e. this class is supposed to be a wrapper around the mutable array?

File: Bearded.Utilities/Noise/NoiseUtils.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
+using System;
+
+namespace Bearded.Utilities.Noise
+{
+    static class NoiseUtils
+    {
+        internal static T[,] GenerateRandomArray<T>(int width, int height, Func<Random, T> generator, int? seed)
```
Comment (amulware) [inline Bearded.Utilities/Noise/NoiseUtils.cs:7]: Instead of injecting a seed, I think it would be better to inject a Random instance. That would allow better control over where numbers come from and can prevent unnecessary allocation.

File: Bearded.Utilities/Noise/PerlinNoise.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    public static class PerlinNoise
+    {
+        // Use uniformly distributed vectors to force better gradient distribution.
+        private const int numVectorSamples = 12;
+        private static readonly Vector2d[] vectorSamples = vectorSamples = Enumerable.Range(0, numVectorSamples)
+            .Select(i => Direction2.FromRadians(MathConstants.TwoPi * i / numVectorSamples).Vector)
+            .Select(vector => new Vector2d(vector.X, vector.Y))
+            .ToArray();
+
+        public static INoiseMap Generate(int numCellsX, int numCellsY, int? seed)
```
Comment (amulware) [inline Bearded.Utilities/Noise/PerlinNoise.cs:17]: It is not clear to me what the numCells parameters do/mean.

Next to that I think this should take a Random, not a seed.

File: Bearded.Utilities/Noise/PerlinNoise.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                throw new ArgumentOutOfRangeException(nameof(numCellsX));
+            }
+            if (numCellsY <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(numCellsY));
+            }
+
+            var gradientArray = NoiseUtils.GenerateRandomArray(
+                // We generate the corners, but to make the noise map wrap along both dimensions, we reuse the same
+                // values as the left and top boundaries, so we don't have to generate the right and bottom boundaries.
```
Comment (amulware) [inline Bearded.Utilities/Noise/PerlinNoise.cs:30]: Similar to above, why do we wannt it to wrap? Should that be the default? Should that be configurable?

**Review by amulware — CHANGES_REQUESTED** [review]

File: Bearded.Utilities/Core/interpolation/Interpolation1.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                {
+                    return to;
+                }
+
+                return (2 * t - 3) * t * t * (from - to) + from;
+            }
+        }
+    }
+
+    public interface IInterpolationMethod1
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Core/interpolation/Interpolation1.cs]: It already is an outer class. I just didn't really want to have a new file for just this one line method. That being said, given the confusion this seems to be leading to, I have extracted these interfaces to separate files.

File: Bearded.Utilities/Noise/INoiseMap.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    public interface INoiseMap
+    {
+        /// <summary>
+        /// Returns the value of the noise map at the given coordinates in the noise map.
+        /// This method only works for values within the [0, 1) x [0, 1) range only (upper bounds exclusive).
+        /// </summary>
+        /// <param name="x"></param>
+        /// <param name="y"></param>
+        /// <returns></returns>
+        public double ValueAt(double x, double y);
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/INoiseMap.cs]: I have never been a fan of using the indexer, because it - in my opinion - removes a bunch of context about what the indexer means. An indexer can be used to access a list, or a map entry, or a continuous value? Having an actual method name here I believe makes it much clearer what the accessor actually means.

I'm aware this is subjective, yet I don't see strong arguments for an indexer apart from succinctness, which I don't believe is necessarily an advantage. I'm happy to be convinced otherwise.

File: Bearded.Utilities/Noise/INoiseMap.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
+namespace Bearded.Utilities.Noise
+{
+    public interface INoiseMap
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/INoiseMap.cs]: Renamed to `ProceduralTexture` since according to [Wikipedia](https://en.wikipedia.org/wiki/Procedural_texture), that is the right name for something like this. Perlin noise specifically is mentioned as an example. I'm open to renaming the entire namespace, or move the non-noise specific procedural texture types to `Core` or their own namespace.

File: Bearded.Utilities/Noise/NoiseMap.cs (hunk 2 above)
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/NoiseMap.cs]: It is, because the centre of the pixel stored at [0, 0] in the array is not actually at texture coordinates (0, 0) so we need to interpolate between the value in [0, 0] and [-1, -1], where the behaviour of [-1, -1] is determined by how the texture wraps.

File: Bearded.Utilities/Noise/NoiseMap.cs (hunk 3 above)
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/NoiseMap.cs]: Imagine a simple noise texture as actual pixels. The array here represents the pixels of that texture. The value you pass in is actually the continuous coordinate on that texture, and there's a good chance that coordinate doesn't exactly align with the centre of a pixel. This is where the interpolation method that gets passed in comes in. If you do bilinear interpolation, you need something to interpolate with. We could interpolate with 0, but this causes artefacts along the outside, which I think is undesirable.

Note that if you use nearest neighbour as interpolation method, this will all still add up to returning the value of pixel [0, 0] in coordinates (0, 0).

> I'm not sure if wrapping is good default behaviour. If anything maybe we have to inject it, if we want to support it?

I agree the wrapping behaviour should be injected, but given the already existent complexity of this PR, has been deferred to future PRs. I think it makes sense to provide a default value though, and I chose repeat in this case (which is in line with the [OpenGL texture default](https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glTexParameter.xml))

File: Bearded.Utilities/Noise/PerlinNoise.cs (hunk 7 above)
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/PerlinNoise.cs:47]: > Similar to above, why do we wannt it to wrap?

By default, yes.

> Should that be the default?

Yes, this sounds the most expected and is in line with OpenGL defaults for traditional textures, and this doesn't immediately break cases that don't expect it.

> Should that be configurable?

Yes, to be added in a future PR.

File: Bearded.Utilities/Noise/PerlinNoise.cs (hunk 6 above)
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/PerlinNoise.cs]: I think you need to know something about how Perlin noise works to understand these parameters. I have added a `<summary>` fragment since I think that is the only way to explain how these parameters map to the underlying algorithm.

File: Bearded.Utilities.Tests/Noise/PerlinNoiseTests.cs [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+using FsCheck.Xunit;
+
+namespace Bearded.Utilities.Tests.Noise
+{
+    public sealed class PerlinNoiseTests
+    {
+        [Property(Arbitrary = new[] { typeof(DoubleGenerators.UnitIntervalUpperBoundExclusive) })]
+        public void GeneratesMapThatThrowsIfCoordinatesTooSmall(int seed, double x, double y)
+        {
+            var map = PerlinNoise.Generate(5, 5, seed);
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities.Tests/Noise/PerlinNoiseTests.cs:14]: As explained in the PR description: the tests om this PR are by no means meant to be exhaustive.

In addition, it is hard to test that these parameters really do something. The randomness involved makes it hard to make hard assertions about the output. There is always a chance that the random output of a 10x10 grid just happens to line up really well with the properties of a 5x5 grid.

We can always generate two maps with the exact same seed with different dimensions and assume they are different in at least one point when converted to a discrete array, but there is no guarantee this is going to be the case for all seeds, and I think even that level of flakiness is not something we want to avoid in our tests.

File: Bearded.Utilities.Tests/Noise/PerlinNoiseTests.cs [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+            Func<double> action = () => map.ValueAt(x - 1, y - 1);
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities.Tests/Noise/PerlinNoiseTests.cs]: Addressed by splitting the tests across dimensions and changing the `1` to `1.0`. The choice for 1 is on purpose as it will exercise edge cases.

File: Bearded.Utilities/Noise/NoiseUtils.cs (hunk 5 above)
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/NoiseUtils.cs:7]: I am happy to be convinced otherwise, but let me explain my reasoning:

One of the properties is that this method returns the exact same result if you give it the same input, i.e. it's deterministic (unless you choose not to pass in a seed). When you replace the seed with a `Random` instance, this because much more nuanced. Sure, if you pass in the "same" `Random`, it is still deterministic. However, there isn't an `Equals` method on `Random`, so the whole concept of having the "same" `Random` is a bit weird. Using a seed makes this much more obviously correct.

Like I said, happy to be convinced otherwise.

**Comment by tomrijnbeek** [comment]
Still awaiting further review I believe.

File: Bearded.Utilities/Noise/PerlinNoise.cs [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+                var yUpper = yLower + 1;
+
+                // Calculate dot products between distance and gradient for each of the grid corners
+                var topLeft = dotProductWithGridDirection(xLower, yUpper, x, y);
+                var topRight = dotProductWithGridDirection(xUpper, yUpper, x, y);
+                var bottomLeft = dotProductWithGridDirection(xLower, yLower, x, y);
+                var bottomRight = dotProductWithGridDirection(xUpper, yLower, x, y);
+
+                // Interpolation weights
+                var tx = Interpolation1.SmoothStep.Interpolate(0, 1, x - xLower);
```
Comment (amulware) [inline Bearded.Utilities/Noise/PerlinNoise.cs:97]: Are te interpolations used the only ones 'valid' for perlin noise?

File: Bearded.Utilities/Noise/UniformNoise.cs [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+namespace Bearded.Utilities.Noise
+{
+    public static class UniformNoise
+    {
+        public static IProceduralTexture Generate(int numCellsX, int numCellsY, int? seed) =>
+            Generate(numCellsX, numCellsY, Interpolation2.Nearest, seed);
+
+        public static IProceduralTexture Generate(int numCellsX, int numCellsY, IInterpolationMethod2 interpolation, int? seed)
+        {
+            if (numCellsX == 0)
```
Comment (amulware) [inline Bearded.Utilities/Noise/UniformNoise.cs:13]: Negative is ok?

**Review by amulware — CHANGES_REQUESTED** [review]
It's been a while since I last took a look, but I think I like it better now. :)

File: Bearded.Utilities.Tests/Core/interpolation/InterpolationMethod1Tests.cs [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+using FluentAssertions;
+using FsCheck.Xunit;
+
+namespace Bearded.Utilities.Tests
+{
+    public abstract class InterpolationMethod1Tests
+    {
+        protected abstract IInterpolationMethod1 Interpolation { get; }
+
+        [Property(Arbitrary = new[] {typeof(DoubleGenerators.NonInfiniteNonNaN)})]
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities.Tests/Core/interpolation/InterpolationMethod1Tests.cs:11]: The latest version of FsCheck does no longer support registering generators in the constructor, because it no longer has static state. I could migrate this now, and it will have to be reverted to update FsCheck, so I'll leave it as is I think.

File: Bearded.Utilities.Tests/Core/interpolation/InterpolationMethod1Tests.cs [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+        public void ReturnsToAtEnd(double from, double to)
+        {
+            Interpolation.Interpolate(from, to, 1).Should().BeApproximately(to, double.Epsilon);
+        }
+
+        [Property(Arbitrary = new[] {typeof(DoubleGenerators.UnitIntervalBoundsInclusive)})]
+        public void ReturnsValuesBetweenFromAndTo(double t)
+        {
+            Interpolation.Interpolate(0, 1, t).Should().BeInRange(0, 1);
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities.Tests/Core/interpolation/InterpolationMethod1Tests.cs:26]: Replaced the 0 and 1 to be injected as well. They still will be between 0 and 1, because hooking specific generators to specific arguments sounds like overkill, but at least it should provide better coverage (including cases for which to < from).

File: Bearded.Utilities.Tests/Core/interpolation/InterpolationMethod2Tests.cs [hunk 17]
```diff
@@ -0,0 +1,45 @@
+using Bearded.Utilities.Tests.Generators;
+using FluentAssertions;
+using FsCheck.Xunit;
+
+namespace Bearded.Utilities.Tests
+{
+    public abstract class InterpolationMethod2Tests
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities.Tests/Core/interpolation/InterpolationMethod2Tests.cs:7]: See other comment.

File: Bearded.Utilities.Tests/Noise/PerlinNoiseTests.cs [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+using Bearded.Utilities.Noise;
+using Bearded.Utilities.Tests.Generators;
+using FluentAssertions;
+using FsCheck.Xunit;
+
+namespace Bearded.Utilities.Tests.Noise
+{
+    public sealed class PerlinNoiseTests
+    {
+        [Property(Arbitrary = new[] { typeof(DoubleGenerators.UnitIntervalUpperBoundExclusive) })]
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities.Tests/Noise/PerlinNoiseTests.cs:11]: See other comment.

File: Bearded.Utilities.Tests/Noise/UniformNoiseTests.cs [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+using Bearded.Utilities.Tests.Generators;
+using FluentAssertions;
+using FsCheck.Xunit;
+
+namespace Bearded.Utilities.Tests.Noise
+{
+    public sealed class UniformNoiseTests
+    {
+        [Property(Arbitrary = new[] { typeof(DoubleGenerators.UnitIntervalUpperBoundExclusive) })]
+        public void GeneratesMapThatThrowsIfXCoordinateTooSmall(int seed, double x, double y)
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities.Tests/Noise/UniformNoiseTests.cs:12]: Added a base class

File: Bearded.Utilities/Core/interpolation/IInterpolationMethod1.cs [hunk 20]
```diff
@@ -0,0 +1,7 @@
+namespace Bearded.Utilities
+{
+    public interface IInterpolationMethod1
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Core/interpolation/IInterpolationMethod1.cs:3]: Renamed them to `IInterpolationMethod1d` and `2d` respectively. I think making this generic makes things much more complicated. I would definitely be up for it, but I don't think there is a good way to make the type restriction "something numeric". I'm open to making it generic, but then do we have separate type parameters for the thing interpreted and the `t` type?

File: Bearded.Utilities/Core/interpolation/Interpolation1.cs [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+            {
+                return t < 0.5 ? from : to;
+            }
+        }
+
+        private sealed class LinearInterpolation : IInterpolationMethod1
+        {
+            public double Interpolate(double from, double to, double t)
+            {
+                if (t <= 0)
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Core/interpolation/Interpolation1.cs:21]: We have benchmarks now, but I don't know if I can be bothered setting those up for interpolations, making this already massive PR even larger.

[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Core/interpolation/Interpolation1.cs:21]: Replaced with Min/Max though.

File: Bearded.Utilities/Noise/PerlinNoise.cs [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+            public PerlinProceduralTexture(Vector2d[,] gradientArray)
+            {
+                this.gradientArray = gradientArray;
+                width = gradientArray.GetLength(0);
+                height = gradientArray.GetLength(1);
+            }
+
+            public double ValueAt(double x, double y)
+            {
+                if (x < 0 || x >= 1)
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/PerlinNoise.cs:70]: Yes, I would like to keep these. Sure, the exceptions will get thrown eventually, but with different numbers and in a different context. I prefer early failure in this case, because it is more obvious what the wrong in parameters are.

File: Bearded.Utilities/Noise/PerlinNoise.cs [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+using System.Linq;
+using Bearded.Utilities.Geometry;
+using OpenTK.Mathematics;
+
+namespace Bearded.Utilities.Noise
+{
+    public static class PerlinNoise
+    {
+        // Use uniformly distributed vectors to force better gradient distribution.
+        private const int numVectorSamples = 12;
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/PerlinNoise.cs:11]: Limiting the randomness... perhaps, yes. I don't know if this is a good value or not, or where I got it from, but the resulting Perlin noise right now looks as expected.

File: Bearded.Utilities/Noise/PerlinNoise.cs [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+                if (x < 0 || x >= 1)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(x));
+                }
+                if (y < 0 || y >= 1)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(y));
+                }
+
+                // First we transform x and y from [0, 1) x [0, 1) to [0, width) x [0, height) for easier math.
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/PerlinNoise.cs:79]: Removed the comments and replaced it with more methods. Not sure if this is clearer, I think I prefer the comments, but unless you tell me otherwise, I'll leave it in this new form.

File: Bearded.Utilities/Noise/PerlinNoise.cs [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+                var bottomLeft = dotProductWithGridDirection(xLower, yLower, x, y);
+                var bottomRight = dotProductWithGridDirection(xUpper, yLower, x, y);
+
+                // Interpolation weights
+                var tx = Interpolation1.SmoothStep.Interpolate(0, 1, x - xLower);
+                // TODO: why is this 1-?
+                var ty = 1 - Interpolation1.SmoothStep.Interpolate(0, 1, y - yLower);
+
+                // Dot products can be between -1 and 1, so we need to normalize before returning.
+                var r = Interpolation2.BiLinear.Interpolate(bottomLeft, bottomRight, topLeft, topRight, tx, ty);
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/PerlinNoise.cs:102]: The normalize was for the line below. That is now its own method.

File: Bearded.Utilities/Noise/PerlinNoise.cs [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+            private readonly int height;
+
+            public PerlinProceduralTexture(Vector2d[,] gradientArray)
+            {
+                this.gradientArray = gradientArray;
+                width = gradientArray.GetLength(0);
+                height = gradientArray.GetLength(1);
+            }
+
+            public double ValueAt(double x, double y)
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/PerlinNoise.cs:68]: Changed.

File: Bearded.Utilities/Noise/ProceduralTexture.cs [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+                height = array.GetLength(1);
+                this.array = new double[width, height];
+                Array.Copy(array, this.array, array.Length);
+
+                this.interpolation = interpolation;
+            }
+
+            public double ValueAt(double x, double y)
+            {
+                if (x < 0 || x >= 1)
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/ProceduralTexture.cs:29]: See my arguments above. I want to keep these for early failure, instead of presenting the failure with something that's really not useful without knowing how it relates back to the original input parameters.

File: Bearded.Utilities/Noise/UniformNoise.cs (hunk 14 above)
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/UniformNoise.cs:13]: Good catch.

File: Bearded.Utilities/Noise/PerlinNoise.cs (hunk 13 above)
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/PerlinNoise.cs:97]: These are part of the definition of Perlin noise, yes.

File: Bearded.Utilities/Core/interpolation/Interpolation1.cs (hunk 21 above)
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Core/interpolation/Interpolation1.cs:21]: The one important disadvantage here is that this means t = 0 means we don't always necessarily return `from` because of double imprecision.

File: Bearded.Utilities/Noise/ProceduralTexture.cs [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+            private readonly int height;
+            private readonly double[,] array;
+            private readonly IInterpolationMethod2 interpolation;
+
+            public ArrayProceduralTexture(double[,] array, IInterpolationMethod2 interpolation)
+            {
+                width = array.GetLength(0);
+                height = array.GetLength(1);
+                this.array = new double[width, height];
+                Array.Copy(array, this.array, array.Length);
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/ProceduralTexture.cs:22]: Dealing with mutability would prevent precalculation, which is something I would like to keep open as an option. Moving from not allowing mutability to allowing mutability is an easier move than the other way around.

File: Bearded.Utilities/Noise/PerlinNoise.cs [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+
+                // Calculate dot products between distance and gradient for each of the grid corners
+                var topLeft = dotProductWithGridDirection(xLower, yUpper, x, y);
+                var topRight = dotProductWithGridDirection(xUpper, yUpper, x, y);
+                var bottomLeft = dotProductWithGridDirection(xLower, yLower, x, y);
+                var bottomRight = dotProductWithGridDirection(xUpper, yLower, x, y);
+
+                // Interpolation weights
+                var tx = Interpolation1.SmoothStep.Interpolate(0, 1, x - xLower);
+                // TODO: why is this 1-?
```
[the remark this answers is not in the document]
Comment (tomrijnbeek) [reply Bearded.Utilities/Noise/PerlinNoise.cs]: 🤷‍♀️

**Comment by tomrijnbeek** [comment]
Ready for another round of review.

File: Bearded.Utilities/Noise/ProceduralTexture.cs [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace Bearded.Utilities.Noise
+{
+    public static class ProceduralTexture
+    {
+        public static IProceduralTexture FromArray(double[,] array, IInterpolationMethod2d interpolation)
+        {
+            var arrowCopy = new double[array.GetLength(0), array.GetLength(1)];
+            Array.Copy(array, arrowCopy, array.Length);
+            return new ArrayProceduralTexture(arrowCopy, interpolation);
```
Comment (amulware) [inline Bearded.Utilities/Noise/ProceduralTexture.cs:11]:
```suggestion
            var arrayCopy = new double[array.GetLength(0), array.GetLength(1)];
            Array.Copy(array, arrayCopy, array.Length);
            return new ArrayProceduralTexture(arrayCopy, interpolation);
```
Typo I assume?

**Review by amulware — APPROVED** [review]
I think you addressed my comments pretty well. Apart from what looks like a typo, go ahead.
